### PR TITLE
Improve container detection

### PR DESCRIPTION
**Pull request by EricYangIBM**
Add logic to detect the presence of container files for both cgroup v1
and v2. If these files are not present, resort to the original method
(which will not work if namespaces are enabled in the container).
Also, the cgroup mount point is renamed as both cgroup v1 and v2 are
conventionally mounted at the same point.

Issue: #1281
Signed-off-by: Eric Yang <eric.yang@ibm.com>

### CODE REVIEW

**Comment by EricYangIBM** [comment]
Tested with java dump:
```
1CICONTINFO    Running in container : TRUE
1CICGRPINFO    JVM support for cgroups enabled : FALSE
```
Second line is `FALSE` since `omrsysinfo_cgroup_is_system_available` has not been updated yet.

fyi @babsingh

**Comment by EricYangIBM** [comment]
I think we may need to change `isRunningInContainer` for cgroup v1 as well. I switched back to v1 and generated a dump with a container created with v2:
```
root@ee74c8987adf:~/hostdir/openj9-openjdk-jdk18# cat /proc/1/cgroup 
12:rdma:/
11:freezer:/
10:blkio:/
9:cpuset:/
8:pids:/
7:hugetlb:/
6:memory:/
5:net_cls,net_prio:/
4:perf_event:/
3:cpu,cpuacct:/
2:devices:/
1:name=systemd:/
0::/
```
All of the cgroups here are root. Should we use the new /.dockerenv check for all cgroup versions? Or maybe we can check the /.dockerenv first and if it's not there check using v1 logic.

I.e.
- check /.dockerenv and/or /run/.containerenv files existence
- if not there, 
-     `isCgroupV1Available`? if yes, use original logic to check /proc/1/cgroup

**Comment by babsingh** [comment]
yes, your proposal in https://github.com/eclipse/omr/pull/6422#issuecomment-1071249090 sounds good.
```
if ((0 == access("/.dockerenv", F_OK)) || (0 == access("/run/.containerenv", F_OK))) {
    *inContainer = TRUE;
} else if (isCgroupV1Available(portLibrary)) {
    <OLD_CODE>
```

File: port/unix/omrsysinfo.c [hunk 1]
```diff
@@ -5884,7 +5906,15 @@ isRunningInContainer(struct OMRPortLibrary *portLibrary, BOOLEAN *inContainer)
 	/* Assume we are not in container */
 	*inContainer = FALSE;
 
-	if (isCgroupV1Available(portLibrary)) {
+	if (isCgroupV2Available(portLibrary)) {
+		/* Check for existence of files that signify running in a container (Docker and Podman respectively). Not completely
+		 * reliable but should work for most cases.
+		 * Ref: https://github.com/systemd/systemd/blob/6604fb0207ee10e8dc05d67f6fe45de0b193b5c4/src/basic/virt.c#L523-L548
```
Comment (babsingh) [inline port/unix/omrsysinfo.c:5912]: The link is not needed in source code.

File: port/unix/omrsysinfo.c [hunk 2]
```diff
@@ -5416,6 +5417,27 @@ isCgroupV1Available(struct OMRPortLibrary *portLibrary)
...
+ *
+ * @return TRUE if cgroup v2 system is available, FALSE otherwise
+ */
+static BOOLEAN
+isCgroupV2Available(struct OMRPortLibrary *portLibrary)
+{
+	BOOLEAN result = FALSE;
+
+	/* If the cgroup.controllers file exists at the root cgroup, then v2 is available */
+	if (0 == access(OMR_CGROUP_V1_MOUNT_POINT "/cgroup.controllers", F_OK)) {
```
Comment (babsingh) [inline port/unix/omrsysinfo.c:5434]: `OMR_CGROUP_V1_MOUNT_POINT` should be renamed if it is used with both cgroup v1 and v2.
```suggestion
	if (0 == access(OMR_CGROUP_MOUNT_POINT "/cgroup.controllers", F_OK)) {
```

File: port/unix/omrsysinfo.c [hunk 3]
```diff
@@ -5884,7 +5906,15 @@ isRunningInContainer(struct OMRPortLibrary *portLibrary, BOOLEAN *inContainer)
 	/* Assume we are not in container */
 	*inContainer = FALSE;
 
-	if (isCgroupV1Available(portLibrary)) {
+	if (isCgroupV2Available(portLibrary)) {
+		/* Check for existence of files that signify running in a container (Docker and Podman respectively). Not completely
+		 * reliable but should work for most cases.
```
Comment (babsingh) [inline port/unix/omrsysinfo.c:5911]:
```suggestion
		 * reliable because these files may not be always available. But, it should work for most cases.
```

**Review by babsingh — CHANGES_REQUESTED** [review]

**Comment by babsingh** [comment]
jenkins build all

**Comment by babsingh** [comment]
jenkins build zos

**Comment by babsingh** [comment]
zos build jobs are failing because the zos machines are inaccessible. restarting the zos jobs does not help since a different zos machine is not chosen by the scheduler. https://ci.eclipse.org/omr/job/PullRequest-zos_390-64/3474/
```
Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to ZISVJD04
10:56:20  	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to ZISVJD04
10:56:20  		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1797)
10:56:20  		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
10:56:20  		at hudson.remoting.Channel.call(Channel.java:1001)
```

**Review by babsingh — APPROVED** [review]

**Comment by EricYangIBM** [comment]
Should this pr be merged before any further changes or should I add changes on top? `omrsysinfo_cgroup_is_system_available` depends on `isCgroupV2Available`

**Comment by EricYangIBM** [comment]
Actually I should move `isCgroupV2Available` to a new PR

File: port/unix/omrsysinfo.c [hunk 4]
```diff
@@ -5884,7 +5884,12 @@ isRunningInContainer(struct OMRPortLibrary *portLibrary, BOOLEAN *inContainer)
 	/* Assume we are not in container */
 	*inContainer = FALSE;
 
-	if (isCgroupV1Available(portLibrary)) {
+	/* Check for existence of files that signify running in a container (Docker and Podman respectively). Not completely
+	 * reliable as these files may not be available, but it should work for most cases.
+	 */
+	if ((0 == access("/.dockerenv", F_OK)) || (0 == access("/run/.containerenv", F_OK))) {
+		*inContainer = TRUE;
+	} else if (isCgroupV1Available(portLibrary)) {
```
Comment (babsingh) [inline port/unix/omrsysinfo.c:5892]: The below check does not apply to cgroup v2. We should avoid reading `/proc/1/cgroup` for cgroup v2. Perf will be better. Is it guaranteed that `isCgroupV1Available` will always fail for cgroup v2?

**Comment by babsingh** [comment]
> Should this pr be merged before any further changes or should I add changes on top? 

Locally, you can work on new changes with changes from this PR. The new work should be published in a new PR. This PR should be merged by the time your next PR is ready. If not merged, we can say `depends on this PR` in the next PR. Once this PR is merged, the other ones can be rebased which will exclude the changes from the previous PRs. 

> Actually I should move isCgroupV2Available to a new PR since it's not needed here

Ok. I felt that the `!isCgroupV2Available` check provided extra safety to avoid reading `/proc/1/cgroup` for cgroup v2. The response to https://github.com/eclipse/omr/pull/6422#discussion_r834315357 should finalize the changes. Once the changes are finalized, I will rerun the PR builds.

File: port/unix/omrsysinfo.c (hunk 4 above)
Comment (EricYangIBM) [reply port/unix/omrsysinfo.c:5892]: For cgroup v2 tmpfs will not be mounted on /sys/fs/cgroup, instead, 'cgroup' filesystem type is mounted there, so I believe our checks for v1 and v2 are mutually exclusive.

**Comment by babsingh** [comment]
Since the code have been updated, can you also update the PR title, description and commit message? 
- [Titile] `Container detection for cgroup v2` -> `Improve container detection`;
- [Body] Mention why `OMR_CGROUP_V1_MOUNT_POINT` is renamed;
- [Body] Remove `logic for detecting if cgroup v2 is enabled`; and
- [Body] Provide more details on the changes to the container detection mechanism.
- [Body + code comment] Reading `/proc/1/cgroup` won't work for container detection in newer cgroup v1.

File: port/unix/omrsysinfo.c (hunk 4 above)
Comment (EricYangIBM) [reply port/unix/omrsysinfo.c:5892]: Why isn't reading /proc/1/cgroup correct for newer v1? From what I have seen cgroup v1 dockers have /proc/1/cgroup outputting nonroot cgroups

Comment (babsingh) [reply port/unix/omrsysinfo.c:5892]: I wanted to refer to your old comment: https://github.com/eclipse/omr/pull/6422#issuecomment-1071249090 where you mentioned `we may need to change isRunningInContainer for cgroup v1 as well. I switched back to v1 ... All of the cgroups here are root.`

Comment (EricYangIBM) [reply port/unix/omrsysinfo.c:5892]: Ok, I believe what's happening there is namespaces being enabled by the v2 container. Namespaces can also be enabled in v1 so I will add a comment explaining that

**Comment by EricYangIBM** [comment]
Yes

**Comment by babsingh** [comment]
Jenkins build all

**Comment by babsingh** [comment]
z/OS failing because of https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/999.

**Comment by babsingh** [comment]
@EricYangIBM Please explicitly acknowledge once all the suggested feedback has been incorporated. Otherwise, it is difficult to interpret if the PR is ready for another review.

**Comment by babsingh** [comment]
@0xdaryl PR builds have passed. This PR should be good to merge.

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**